### ISSUE TITLE
got "Maximum call stack size exceeded" with compileAsync

### ISSUE THREAD

**Issue by goto100** [ajv-validator/ajv#220]
The issue happens when $ref's hash has match the original schema, here is the code:

```
const Ajv = require('ajv');

const schema = {
  properties: {
    name: {
      $ref: 'Something#/properties/name',
    },
  },
};

const asyncSchema = {
  properties: {
    name: {
      type: 'string',
    },
  },
};

const data = {
  name: 'haha',
};

var ajv = new Ajv({ loadSchema: loadSchema });

ajv.compileAsync(schema, function (err, validate) {
    if (err) {
      console.log(err); // got Maximum call stack size exceeded
      return;
    }
    var valid = validate(data);
    console.log(valid);
});

function loadSchema(uri, callback) {
  setTimeout(() => {
    callback(null, asyncSchema);
  }, 100);
}
```

I debugged Ajv's code, the order of ajv try to resolve a $ref is "added schemas > schema it self > loadSchema function", I think the order should be changed to "added schemas > loadSchema function > schema it self"

**Comment by epoberezkin** [comment]
Interesting. If you replace "properties" in asyncSchema and in $ref with "definitions" (or anything else) it works: https://tonicdev.com/esp/5772c6738dc3051200da26e9

**Comment by epoberezkin** [comment]
And using different property names in schemas also fixes the issue...

**Comment by epoberezkin** [comment]
So the real issue has nothing to do with async loading - it happens because ref resolution doesn't fail when it should (because referenced schema is missing): https://tonicdev.com/esp/5772cc02b5157813007e909a

**Comment by epoberezkin** [comment]
fixed in 4.1.5

**Comment by goto100** [comment]
So efficient! Thank you!
